KaleidoscopeLang/Expression.swift

```diff
@@ -21,7 +21,7 @@ extension Expression where Self : Equatable {
         return false
     }
 }
-public func == (left: Expression, right: Expression) -> Bool {
+public func == <E: Expression> (left: E, right: E) -> Bool {
```

Comment:
Constraining thus avoids an infinite loop between `.equals` and `==`. It’s unclear to me _why_ that is so, however—I’d have hoped the use of `==` in the protocol extension would have selected the `Equatable` version!